Improvement: MXP text effects: bold, underline, italics, strikethrough

eowmob opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
MXP text effects (incl. all aliases) are now rendered.
Colour and text effects in open mode auto-close at end of line 
#### Motivation for adding to Mudlet
Improve MXP compliance
#### Other info (issues closed, discussion etc)
[MXP enhancements, was PR #4200](https://forums.mudlet.org/viewtopic.php?f=7&t=23206)

eowmob committed:
MXPFormatting is now applied

Formatting tags can be nested and are rendered accordingly.
They are closed automatically according to MXP spec.

eowmob committed:
Merge branch 'Mudlet:development' into WorkingSimpleMXPFormatting

eowmob committed:
Fixed typo in test\CMakeLists.txt for FormattingTagsTest

eowmob committed:
Merge branch 'Mudlet:development' into WorkingSimpleMXPFormatting

vadi2 commented:
What's a game where I can test the before & after? I know you mentioned it once but I can't find it again.

vadi2 commented:
/create links

vadi2 commented on src/TMxpClient.h:
```diff
@@ -65,6 +65,14 @@ class TMxpClient
     virtual void setUnderline(bool val) = 0;
+    virtual void setStrikeOut(bool val) = 0;
+
+    virtual bool isBold() = 0;
```
Mind renaming this to just `bold()` and others too? It would be consistent with how we use `mBold` in TBuffer and with Qt's naming philosophy too.

eowmob replied:
@first: Sorry, I didn't notice your comments during the last month (sth with my notification settings), so I missed this completely, thks for the reminder assignment. I'm in the process of making a new commit that deals with all of them, except maybe (*) below where I might have a question. Should be done in 1-2h.

eowmob replied:
regarding isBold() -> bold() and friends, ok sure, why not.

vadi2 commented on src/TMxpMudlet.h:
```diff
@@ -36,14 +36,16 @@ class TMediaData;
 class TMxpMudlet : public TMxpClient
 {
+    // Count how many of this format have been stacked/applied on top of each other
+    unsigned int boldCtr = 0;
```
Would prefer `boldCounter` for this & others, makes it just a bit nicer to read.

eowmob replied:
ok, changing it right now.

vadi2 commented on src/TMxpMudlet.cpp:
```diff
@@ -114,3 +114,55 @@ TLinkStore& TMxpMudlet::getLinkStore()
+{
+    if (bold) {
+        boldCtr++;
+    } else if (boldCtr > 0) {
```
Nice construct!

eowmob replied:
Thanks, I mainly wanted to keep the existing object interface for compatibility.

vadi2 commented on src/TMxpMudlet.cpp:
```diff
@@ -114,3 +114,55 @@ TLinkStore& TMxpMudlet::getLinkStore()
+        fgColors.pop_back();
+
+    while (!bgColors.isEmpty())
+        bgColors.pop_back();
```
```suggestion
    while (!fgColors.isEmpty()) {
        fgColors.pop_back();
    }

    while (!bgColors.isEmpty()) {
        bgColors.pop_back();
    }
```

Better to be explicit about it, we prefer it this way in the codebase.

vadi2 commented on src/TMxpMudlet.cpp:
```diff
@@ -114,3 +114,55 @@ TLinkStore& TMxpMudlet::getLinkStore()
+    }
+}
+
+// reset text Properties (from open tags) at end of line
```
Could you update the comment to explain why it just removes one element from the vector and doesn't clear it entirely, as one would intuit?

eowmob replied:
Ok, I added a comment:
    // for the next two, we can assume both lists are usually empty (in case of properly
    // balanced MXP tags, and if not, they'll only contain very few entries. Thus it seems
    // sensible to check first, and then just remove all of the few nodes rather than
    // removing the whole list first and then create a new one.
 and wrote it that way since I think it is the most ressource friendly way to do it in the vast majority of cases. But I think you might prefer something like:
 
    if (!fgColors.isEmpty()) {
        fgColors = QList<QColor>();
    }
    
 Just drop me a note if I should go for that one. In principle, all of these are fine for me. Oh, this is the (*) question I referred to from above.

vadi2 replied:
This works ok!

vadi2 commented on src/TMxpProcessor.cpp:
```diff
@@ -77,13 +77,20 @@ bool TMxpProcessor::setMode(int modeCode)
         break;
     case 1: // secure line (until next newline) all tags and commands in MXP are allowed within the line.  When a newline is received from the MUD, the mode reverts back to the Default mode.
+        // When the mode is changed from OPEN mode to any other mode, any unclosed OPEN tags are automatically closed.
+        if (mMXP_MODE == MXP_MODE_OPEN)
```
Here & in other places on the one-line braces. Just learning from Apple's goto fail, and the tax isn't that high once you get used to it

eowmob commented:
> What's a game where I can test the before & after? I know you mentioned it once but I can't find it again.

Well, you can use aldebaran-mud.de port 2000 . I can't tell you to what other muds use MXP and to which extent, sorry.
You can just login as guest (there can be many guest chars online) or create a char. All will spawn in the same room.

Now, admittedly the game does not use MXP text styles in its 'system' messages (it uses vt100/ANSI as a more common denominator), but you can say something like: say ^This is &lt;I>italics&lt;/I>.
Just don't forget to prefix the text with a ^ character, otherwise the "&lt;" character is escaped and the MXP tags do not make it back to Mudlet. Maybe underline or &lt;U> is not such a good test here, as the said text is also an mxp link and underlined anyway.

To test underline, you can create a bulletin note (next board from the church where you login is s,e,e,e,s). For example:

note test
^This is &lt;U>underlined
^This is &lt;U>not&lt;/U> see?
.
read (whatever number the new note got, check with: examine board... [the clickable MXP links in the board don't work.. another, unrelated, MXP issue with mudlet: If I remember right it is about interpolation of entities in elements])
Again, do not forget to prefix each line with a ^ character, otherwise the &lt; character is escaped by the mud and no MXP tag send to Mudlet.


BTW, when you login two times (using the released mudlet 4.17.2), for example with two guest chars (I figured you'll have to create two profiles in mudlet to connect to the same game), you can also demonstrate the 'black on black attack' I refer to so often when trying to explain to people why there is an open and a closed mode in MXP:

guest1 & 2 are in the same room.
guest1:   say ^&lt;C fore=black back=black>this is black
Both (actually all MXP capable clients in the same room) are blinded.
guest1 leaves the room, say he leaves the church to the south (he needs to remember there was an exit to the south). guest2 does not notice it, he is blinded.
guest1 does: say ^&lt;/C> to restore his vision.
As he is no longer in the same room as guest2, the vision of guest2 is not restored.
guest2, if he has no clue what happened, is now totally lost. Even disconnecting from the game and reconnecting does not help, as the color is not reset upon reconnecting, and the mud cannot reset it at login with ESC[3z. He actually has to close and reopen mudlet. Actually, even while online, guest2 could enter (blindly) "cls sane" or "fixme" . But as mudlet does not respond to the MXP reset sequence ESC[3z, it won't help him.

This will not work with the patch, Mudlet will simply revert to the default colour after writing "this is black" in black on black.

Thanks for your effort and time.

eowmob commented:
Just editted the previous note as the browser interpreted all the embedded MXP tags

eowmob committed:
Merge branch 'Mudlet:development' into WorkingSimpleMXPFormatting

vadi2 reviewed (approved):
Working great, just have a couple of code styling improvements

eowmob committed:
Merge remote-tracking branch 'upstream/development' into WorkingSimpleMXPFormatting

eowmob committed:
merge qtkeychain submodule commit.

eowmob committed:
Changed naming of current mxp style functions

isBold()->bold()
isStrikeOut()->strikeOut()
isUnderline()->underline()
isItalic()->italic()

eowmob committed:
Changed variable names _x_Ctr to _x_Counter, uncommented property checks in Custom Element tests

eowmob committed:
Minor changes according to apple's goto fail.

eowmob commented:
@vadi2 Ok, I think I incorporated all your comments. Please check for the comment on "while (!fgColors.isEmpty()) {}" above.

I hope the 'merge qtkeychain submodule commit.' did not break anything, I thought it's cleaner to base it on the current development tree.

vadi2 commented:
It all looks good to me, though could you revert the qtkeychain changes? It should be set to this commit - https://github.com/Mudlet/qtkeychain/tree/ea7e748a44d0f858b756b78cfa7272075949d319

eowmob committed:
Revert "merge qtkeychain submodule commit."

This reverts commit 038c1e7094ef3ad47363fb291799ae7bdcf6c020.

eowmob commented:
@vadi2 Alas, now Appveyor failed, IMHO due to some internal, unrelated, issue.  Looks like a race condition. Do you have a chance to rerun the tests?

I think I managed Github Desktop to not insist on the submodule change anymore, I've no idea what made it think the file should be changed. Things like this happen from time to time in Github Desktop.
I should be able to commit/revert some tiny comment change to trigger a rerun and possible automerge, but I wonder what is the best option here.
